src/main/java/net/wasdev/wlp/common/plugins/util/DevUtil.java

```diff
@@ -0,0 +1,956 @@
+                    sourceDirRegistered = false;
+                }
+
+                // check if testSourceDirectory has been added
+                if (!testSourceDirRegistered && this.testSourceDirectory.exists()) {
+                    compile(this.testSourceDirectory);
+                    registerAll(this.testSourceDirectory.toPath(), testSrcPath, watcher);
+                    debug("Registering Java test directory: " + this.testSourceDirectory);
+                    runTestThread(false, executor, -1, false, false);
+                    testSourceDirRegistered = true;
+                } else if (testSourceDirRegistered && !this.testSourceDirectory.exists()) {
+                    cleanTargetDir(testOutputDirectory);
+                    testSourceDirRegistered = false;
+                }
+                
+                // check if configDirectory has been added
+                if (!configDirRegistered && this.configDirectory.exists()){
+                    configDirRegistered = true;
+                    debug("Configuration directory has been added: " + this.configDirectory);
+                    info("The server configuration directory " + configDirectory + " has been added. Restart liberty:dev mode for it to take effect.");
+                }
+                
+                // check if configFile has been added
+                if (!configFileRegistered && configFile.exists()){
+                    configFileRegistered = true;
+                    debug("Configuration file has been added: " + configFile);
+                    info("The server configuration file " + configFile + " has been added. Restart liberty:dev mode for it to take effect.");
+                }
+                
+                try {
+                    final WatchKey wk = watcher.poll(1, TimeUnit.SECONDS);
+                    for (WatchEvent<?> event : wk.pollEvents()) {
+                        final Path changed = (Path) event.context();
+
+                        final Watchable watchable = wk.watchable();
+                        final Path directory = (Path) watchable;
+                        debug("Processing events for watched directory: " + directory);
+
+                        File fileChanged = new File(directory.toString(), changed.toString());
+                        debug("Changed: " + changed + "; " + event.kind());
+
+                        // resource file check
+                        File resourceParent = null;
+                        for (File resourceDir : resourceDirs) {
+                            if (directory.startsWith(resourceDir.toPath())) {
+                                resourceParent = resourceDir;
+                            }
+                        }
+                        
+                        int numApplicationUpdatedMessages = countApplicationUpdatedMessages();
+
+                        // src/main/java directory
+                        if (directory.startsWith(this.sourceDirectory.toPath())) {
```

Comment:
When doing these string comparisons here, I think it's more robust if we get the canonical path of the sourceDirectory, etc.

Reply:
fixed with https://github.com/WASdev/ci.common/pull/48/commits/83e2391e82f984eb04367b51310869a7c9853607